### ISSUE TITLE
Not recognizing paths with left square bracket

### ISSUE THREAD

**Issue by serjflint** [pytest-dev/pytest#10329]
- [x] a detailed description of the bug or problem you are having
- [ ] output of `pip list` from the virtual environment you are using
- [x] pytest and operating system versions
- [x] minimal example if possible

`pytest` skips everything after a left square bracket if provided with a path to a folder

```
pytest
============================= test session starts ==============================
platform linux -- Python 3.10.6, pytest-7.1.3, pluggy-1.0.0
rootdir: /home/user/pytest_test
plugins: cov-3.0.0, timeout-2.1.0, requests-mock-1.10.0, anyio-3.6.1, mock-3.8.2
collected 2 items                                                              

simple/d1/test_d1.py .                                                   [ 50%]
simple[hard]/d2/test_d2.py .                                             [100%]

============================== 2 passed in 0.01s ===============================
```

```
pytest simple\[hard\]/
============================= test session starts ==============================
platform linux -- Python 3.10.6, pytest-7.1.3, pluggy-1.0.0
rootdir: /home/user/pytest_test
plugins: cov-3.0.0, timeout-2.1.0, requests-mock-1.10.0, anyio-3.6.1, mock-3.8.2
collected 1 item                                                               

simple/d1/test_d1.py .                                                   [100%]

============================== 1 passed in 0.01s ===============================
```
Check out the provided path in the second block and the path `pytest` tests

**Comment by KingMichaelPark** [comment]
Have you tried putting the second command in " "

**Comment by serjflint** [comment]
> Have you tried putting the second command in " "

I tried now and updated the question

**Comment by nnrepos** [comment]
@KingMichaelPark 
here's the culprit: https://github.com/pytest-dev/pytest/pull/9642
it seems like square brackets are considered part of pytest's syntax.
not sure if this is a bug or lack of documentation/warning then.
if you're interested in a fix, please let me know 😄

**Comment by KingMichaelPark** [comment]
I think you should rename your directory 😉

**Comment by nnrepos** [comment]
> I think you should rename your directory 😉

this is clearly not a sufficient solution. what about projects which use brackets in hundreds, thousands of directories?
something like this should, at the very least:
* be well-documented
* raise a warning

@nicoddemus, could you share your opinion please?

**Comment by RonnyPfannschmidt** [comment]
Brackets are not supported as part of python module names, so unless extra contex is provided i consider it a no go

**Comment by nnrepos** [comment]
> Brackets are not supported as part of python module names, so unless extra contex is provided i consider it a no go

according to [the docs](https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html):
```
If no arguments are specified then collection starts from testpaths (if configured) or the current directory.
Alternatively, command line arguments can be used in any combination of directories, file names or node ids.
```
meaning that pytest also supports test **directories**, not just test **modules**.

regardless of that, shouldn't the code be able to identify and warn about these "illegal" directory names?
i think it would help users understand what is going on, and rename their directories.

**Comment by RonnyPfannschmidt** [comment]
Some type of warning seems sensible

As far as I can tell in the last 15 years this is the first time that brackets in folder names came up as a request

**Comment by KingMichaelPark** [comment]
Yeah I have never seen them, what's the use case for hundreds of thousands of directories with brackets in them (completely out of curiosity not a criticism)

**Comment by nnrepos** [comment]
> Some type of warning seems sensible
> 
> As far as I can tell in the last 15 years this is the first time that brackets in folder names came up as a request

that's probably because the PR that broke them was pushed this february, and not everyone is using the latest pytest version.
if you would like me to add a warning in a new PR, please let me know.
thanks :)

**Comment by nnrepos** [comment]
> Yeah I have never seen them, what's the use case for hundreds of thousands of directories with brackets in them (completely out of curiosity not a criticism)

i honestly don't know, i just know it's possible 😅

**Comment by asottile** [comment]
I'd rather just say "play stupid games win stupid prizes" -- putting square brackets in directory names imo clearly falls into that category

**Comment by MaggiR** [comment]
Hi there!

I also just ran into that issue on my Windows machine. It took me 3 hours to determine that the root cause comes from pytest's inability to handle brackets in the path. I am using PyCharm, which passes absolute (full) paths to `pytest` as arguments. As a consequence, **if any of the test root's parent directories (even outside the project) have "[" or "]" in their name, pytest breaks.**

A simple warning would be very helpful here. 🙂
